lib/inventory/inventory.go

```diff
@@ -614,7 +636,35 @@ func (h *upstreamHandle) Ping(ctx context.Context, id uint64) (d time.Duration,
 
 	select {
 	case rsp := <-rspC:
-		return rsp.d, rsp.err
+		return rsp.reqDuration, rsp.err
+	case <-h.Done():
+		return 0, trace.Errorf("failed to recv upstream pong (stream closed)")
+	case <-ctx.Done():
+		return 0, trace.Errorf("failed to recv upstream ping: %v", ctx.Err())
+	}
+}
+
+// TimeReconciliation makes ping request to compare time difference between upstream and downstream.
+func (h *upstreamHandle) TimeReconciliation(ctx context.Context, id uint64) (d time.Duration, err error) {
+	rspC := make(chan pingResponse, 1)
+	select {
+	case h.pingC <- pingRequest{rspC: rspC, id: id}:
+	case <-h.Done():
+		return 0, trace.Errorf("failed to send downstream ping (stream closed)")
+	case <-ctx.Done():
+		return 0, trace.Errorf("failed to send downstream ping: %v", ctx.Err())
+	}
+
+	select {
+	case rsp := <-rspC:
+		switch {
+		case rsp.clockDiff == 0:
+			return 0, nil
```

Comment:
Instead of returning zero here, let's return an error that indicates it's not supported as clock diffs can also be zero. 

The caller can then errors.Is to ignore the scenario